### PR TITLE
Spans Table View 

### PR DESCRIPTION
**Pull request by vvvprabhakar**
Signed-off-by: vvvprabhakar <vvvprabhakar@gmail.com>

**Why trace span table?** Issue(https://github.com/jaegertracing/jaeger-ui/issues/690)
A trace with many spans trace (~20K) spans can be broken down into pieces and analyzed. Have a link with `traceTimeline View`  as well to link both views and get the analysis of trace 
Gives a detailed view of a service or an operation and its impact on a trace,

**What can we achieve?**

- Filter Spans with multiple serviceName values

- Filter Spans with multiple OperationName values

- Filter with both ServiceName and OperationName 

- Sort based on duration and startTime 

- Each row in the table has spanID link which will take the user to that particular span in `traceTimelineView` 

- Either search with text on serviceName and OperatioName Columns

- Pagination for the spans

**What can we improve from here?** 

- We can add SelfTime in the table as a column and make it sortable

**Dependencies Used** 

- Used antd for table 

**Screenshot 2021-06-23 at 15 51 11
Screenshot 2021-06-23 at 15 51 20
Screenshot 2021-06-23 at 15 51 36
Screenshot 2021-06-23 at 15 52 11
**

### CODE REVIEW

**Comment by jpkrohling** [comment]
The tests seem to be failing. Are you able to sort them out?

**Comment by vvvprabhakar** [comment]
@jpkrohling  Need your help in fixing this unit case(1 case)

**Comment by rubenvp8510** [comment]
Hi @vvvprabhakar  there is a small bug on the `AltViewOptions`  tests, in the `'track dropdown menu'`  case:

Check this commit I did on my own fork:

https://github.com/rubenvp8510/jaeger-ui/commit/09664d94bd5f7d39b23fb384366e4965262ba5d0

It was a coincidence that in the past,` link2.length === links2.length` but now that you added another entry the bug comes up and the test start to fail.

With this change your tests should pass, at least it pass on my local.  :)

**Comment by vvvprabhakar** [comment]
@rubenvp8510  Thanks a lot man

**Comment by vvvprabhakar** [comment]
@rubenvp8510 can you review once 
I have optimised the filtering logic also

**Comment by vvvprabhakar** [comment]
Once you get time  can you please review this

**Comment by jpkrohling** [comment]
ping @rubenvp8510

**Comment by rubenvp8510** [comment]
I'll review this tomorrow morning, overall looks good but I want to see it on detail

File: packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.tsx [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+const Option = Select.Option;
+
+function getNestedProperty(path: string, span: any): string {
+  return path.split('.').reduce((prev, curr) => {
+    return prev ? prev[curr] : null;
+  }, span);
+}
+
+function timeConversion(microseconds: number) {
```
Comment (rubenvp8510) [inline packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.tsx:33]: I think we can move this to utils (outside the component) and write a couple of unit tests.

File: packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.tsx [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+};
+
+export default class TraceSpanView extends Component<Props, State> {
+  constructor(props: Props, state: State) {
+    super(props, state);
+    const serviceNamesList = new Set<string>();
+    const operationNamesList = new Set<string>();
+    const serviceNameOperationsMap = new Map<string, string[]>();
+
+    this.props.trace.spans.map(span => {
```
Comment (rubenvp8510) [inline packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.tsx:82]: Why are you using a map here? I don't see where you use the mapped values.

File: packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.tsx [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+      serviceNamesList.add(span.process.serviceName);
+      operationNamesList.add(span.operationName);
+      const operationNames = serviceNameOperationsMap.get(span.process.serviceName) || [];
+      operationNames.push(span.operationName);
+      serviceNameOperationsMap.set(span.process.serviceName, operationNames);
+      return { serviceNamesList, operationNamesList };
```
Comment (rubenvp8510) [inline packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.tsx:88]: serviceNamesList

**Comment by rubenvp8510** [comment]
Can we add a couple of tests to the new views? at least basic testing?

File: packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.tsx (hunk 1 above)
Comment (vvvprabhakar) [reply packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.tsx:33]: Ok will take this

File: packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.tsx (hunk 2 above)
Comment (vvvprabhakar) [reply packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.tsx:82]: Removed this

File: packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.tsx (hunk 3 above)
Comment (vvvprabhakar) [reply packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.tsx:88]: We are not using it removed

**Comment by vvvprabhakar** [comment]
@rubenvp8510  made the changes you proposed 
Can you just reverify them once you get the time?
Thanks in Adavance

**Comment by rubenvp8510** [comment]
Thanks for doing the changes! 

I'll review tomorrow morning!

**Comment by vvvprabhakar** [comment]
Did you get a chance to see this?

File: packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.test.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    expect(wrapper.find('colgroup').length).toBe(1);
+    expect(wrapper.find('Pagination').length).toBe(2);
+
+    expect(wrapper.find('Button').length).toBe(1);
+    expect(wrapper.find('.ant-form-item-control').length).toBe(3);
+  });
+  it('Should change value when onChange was called', () => {
+    const event = ['service2'];
+    wrapper = shallow(<TraceSpanView {...defaultProps} />);
+    // wrapper.state('filtered.process.serviceName', ['service1', 'service2']);
```
Comment (rubenvp8510) [inline packages/jaeger-ui/src/components/TracePage/TraceSpanView/index.test.js:52]: Remove this comment please.

**Comment by rubenvp8510** [comment]
Small comment, overall looks good. I just need to do a quick smoke test and will merge.

**Comment by vvvprabhakar** [comment]
@rubenvp8510  Removed the line and pushed it 
Thanks for taking time and reviewing it

**Comment by vvvprabhakar** [comment]
@rubenvp8510  did you get time to see this ?

**Review by rubenvp8510 — APPROVED** [review]

**Comment by rubenvp8510** [comment]
I've already approved but it will be good if we can increase the coverage a little bit. may be we can write tests for https://github.com/jaegertracing/jaeger-ui/pull/781/files#diff-00ec9fc101b0b3db820b0352d451f88833915d001216cf78d40ac1178a65be28R148

**Comment by vvvprabhakar** [comment]
Thanks for reviewing it
@rubenvp8510  updated the test cases also

**Comment by vvvprabhakar** [comment]
Coverage was 0.05% less, so covered it, when you have time just check this @rubenvp8510 
Thanks

**Review by rubenvp8510 — APPROVED** [review]
Thanks!

**Pull request merged**